api/src/test/java/org/pmiops/workbench/workspaces/WorkspacesControllerTest.java

```diff
@@ -828,68 +832,102 @@ public void testListForApproval() throws Exception {
-    Workspace workspace2 =
+    final Workspace clonedWorkspace =
         workspacesController
-            .cloneWorkspace(workspace.getNamespace(), workspace.getId(), req)
+            .cloneWorkspace(originalWorkspace.getNamespace(), originalWorkspace.getId(), req)
             .getBody()
             .getWorkspace();
     verify(mockWorkspaceAuditAdapter)
-        .fireDuplicateAction(any(DbWorkspace.class), any(DbWorkspace.class));
+        .fireDuplicateAction(anyLong(), anyLong(), any(Workspace.class));
 
     // Stub out the FC service getWorkspace, since that's called by workspacesController.
-    stubGetWorkspace(clonedWorkspace, WorkspaceAccessLevel.WRITER);
+    stubGetWorkspace(clonedFirecloudWorkspace, WorkspaceAccessLevel.WRITER);
+    final Workspace retrievedWorkspace =
+        workspacesController
+            .getWorkspace(clonedWorkspace.getNamespace(), clonedWorkspace.getId())
+            .getBody()
+            .getWorkspace();
+
+    // Hack so lists can be compared in isEqualTo regardless of order. Order doesn't matter
+    // semantically, but I don't want to go down the rabbit hole of an out-of-class equality
+    // method or custom assertion here (which would soon go out of date).
+    sortPopulationDetails(clonedWorkspace.getResearchPurpose());
+    sortPopulationDetails(retrievedWorkspace.getResearchPurpose());
+    sortPopulationDetails(modPurpose);
+
     assertWithMessage("get and clone responses are inconsistent")
-        .that(workspace2)
-        .isEqualTo(
-            workspacesController
-                .getWorkspace(workspace2.getNamespace(), workspace2.getId())
-                .getBody()
-                .getWorkspace());
+        .that(clonedWorkspace)
+        .isEqualTo(retrievedWorkspace);
+
+    assertThat(clonedWorkspace.getName()).isEqualTo(modWorkspace.getName());
+    assertThat(clonedWorkspace.getNamespace()).isEqualTo(modWorkspace.getNamespace());
+    assertThat(clonedWorkspace.getResearchPurpose()).isEqualTo(modPurpose);
+  }
+
+  private void sortPopulationDetails(ResearchPurpose researchPurpose) {
+    final List<SpecificPopulationEnum> populateionDetailsSorted =
+        researchPurpose.getPopulationDetails().stream().sorted().collect(Collectors.toList());
+    researchPurpose.setPopulationDetails(populateionDetailsSorted);
+  }
 
-    assertThat(workspace2.getName()).isEqualTo(modWorkspace.getName());
-    assertThat(workspace2.getNamespace()).isEqualTo(modWorkspace.getNamespace());
-    assertThat(workspace2.getResearchPurpose()).isEqualTo(modPurpose);
+  private UserRole createUserToShareWith(
+      ShareWorkspaceRequest shareWorkspaceRequest,
+      String email,
+      WorkspaceAccessLevel workspaceAccessLevel) {
+    final UserRole userRole = new UserRole();
+    userRole.setEmail(email);
+    userRole.setRole(workspaceAccessLevel);
+    shareWorkspaceRequest.addItemsItem(userRole);
+    return userRole;
+  }
+
+  private DbUser createAndSaveUser(String email, long userId) {
```

Comment:
Do we have to pass in an ID with our create method? I know we do this here and there in our test code but I think there's a better solution.

We can just grab the generated ID from the saved object if we need it for comparison later on in the test.

Manually setting the ID slightly increases the likelihood of a flaky test if there are ID conflicts between tests.

Reply:
That's consistent with what we do elsewhere (relying on the Dao to pick the id). It weirds me out that a controller test has a live DAO at all though. I feel like that layer should ideally be mocked, and maybe the service layer too. But the WorkspacesController is special in that regard, until we get around to refactoring it.

Reply:
I think you're right about the DAO being reset but I've run into some conflicts even within a single test while wrangling through stuff in `WorkspaceControllerTest`. Refactoring and moving stuff around broke stuff since models no longer had a unique ID.